packages/itwinui-react/src/core/Table/actionHandlers/selectHandler.ts

```diff
@@ -38,20 +39,23 @@ const onSelectHandler = <T extends Record<string, unknown>>(
     }
 
     let isAllSubSelected = true;
-    row.initialSubRows.forEach((subRow) => {
-      const result = handleRow(subRow);
-      if (!result) {
-        isAllSubSelected = false;
-      }
-    });
+    if (row.subRows[0]?.original[iuiId as any] === undefined) {
+      row.initialSubRows.forEach((subRow) => {
+        const result = handleRow(subRow);
+        if (!result) {
+          isAllSubSelected = false;
+        }
+      });
+    }
 
     // If `selectSubRows` is false, then no need to select sub-rows and just check current selection state.
     // If a row doesn't have sub-rows then check its selection state.
     // If it has sub-rows then check whether all of them are selected.
     if (
-      (!instance.selectSubRows && newState.selectedRowIds[row.id]) ||
-      (!row.initialSubRows.length && newState.selectedRowIds[row.id]) ||
-      (row.initialSubRows.length && isAllSubSelected)
+      newState.selectedRowIds[row.id] &&
+      (!instance.selectSubRows ||
+        !row.initialSubRows.length ||
+        (row.initialSubRows.length && isAllSubSelected))
```

Comment:
I was thinking if we could simplify the if statement since they all needed the `newState.selectedRowIds[row.id]` in every `||` check so I just did so to avoid duplication.

Reply:
I have updated the comment as suggested :)